app/src/main/java/com/firebase/uidemo/database/realtime/RealtimeDbChatActivity.java

```diff
@@ -0,0 +1,177 @@
+ * the {@link FirebaseRecyclerAdapter} to build a simple chat app.
+ * <p>
+ * For a general intro to the RecyclerView, see <a href="https://developer.android.com/training/material/lists-cards.html">Creating
+ * Lists</a>.
+ */
+public abstract class RealtimeDbChatActivity extends LifecycleActivity
+        implements FirebaseAuth.AuthStateListener {
+    private static final String TAG = "RealtimeDatabaseDemo";
+
+    /**
+     * Get the last 50 chat messages.
+     */
+    protected static final Query sChatQuery =
+            FirebaseDatabase.getInstance().getReference().child("chats").limitToLast(50);
+
+    @BindView(R.id.messagesList)
+    RecyclerView mRecyclerView;
+
+    @BindView(R.id.sendButton)
+    Button mSendButton;
+
+    @BindView(R.id.messageEdit)
+    EditText mMessageEdit;
+
+    @BindView(R.id.emptyTextView)
+    TextView mEmptyListMessage;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_chat);
+        ButterKnife.bind(this);
+
+        mRecyclerView.setHasFixedSize(true);
+        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
+
+        ImeHelper.setImeOnDoneListener(mMessageEdit, new ImeHelper.DonePressedListener() {
+            @Override
+            public void onDonePressed() {
+                onSendClick();
+            }
+        });
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        if (isSignedIn()) { attachRecyclerViewAdapter(); }
+        FirebaseAuth.getInstance().addAuthStateListener(this);
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        FirebaseAuth.getInstance().removeAuthStateListener(this);
+    }
+
+    @Override
+    public void onAuthStateChanged(@NonNull FirebaseAuth auth) {
+        mSendButton.setEnabled(isSignedIn());
+        mMessageEdit.setEnabled(isSignedIn());
+
+        if (isSignedIn()) {
+            attachRecyclerViewAdapter();
+        } else {
+            Toast.makeText(this, R.string.signing_in, Toast.LENGTH_SHORT).show();
+            auth.signInAnonymously()
+                    .addOnSuccessListener(this, new OnSuccessListener<AuthResult>() {
+                        @Override
+                        public void onSuccess(AuthResult result) {
+                            attachRecyclerViewAdapter();
+                        }
+                    })
+                    .addOnCompleteListener(new SignInResultNotifier(this));
+        }
+    }
+
+    private boolean isSignedIn() {
+        return FirebaseAuth.getInstance().getCurrentUser() != null;
+    }
+
+    private void attachRecyclerViewAdapter() {
+        final RecyclerView.Adapter adapter = newAdapter();
+
+        // Scroll to bottom on new messages
+        adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
```

Comment:
Can we replace the non-scrolling pieces of this by watching onDataChanged?

Reply:
Yep, old code from the abstraction.